core/promise_ring.rs

```diff
@@ -0,0 +1,73 @@
+use crate::PromiseId;
+use std::collections::btree_map::BTreeMap;
+
+type PromiseResolver = v8::Global<v8::PromiseResolver>;
+const RING_SIZE: usize = 4 * 1024;
+
+// NOTE: could generalize to a "RingMap"
+pub struct PromiseRing {
+  len: u32,
+  cursor: u32,
+  // TODO(@AaronO): Maybe optimize by tracking existence with a bitset
+  // would reduce mem usage from 2*usize*RING_SIZE to (1+1/8)*usize*RING_SIZE
+  // with the current approach the Option tag "wastes" 32kb per isolate
+  ring: Vec<Option<PromiseResolver>>,
+  map: BTreeMap<PromiseId, PromiseResolver>,
+}
+
+impl PromiseRing {
+  pub(crate) fn new() -> Self {
+    Self {
+      len: 0,
+      cursor: 0,
+      ring: vec![None; RING_SIZE],
+      map: BTreeMap::new(),
+    }
+  }
+
+  // TODO(@AaronO): decide if this is useful (but it should match length of futures stream so ...)
+  // pub(crate) fn len(&self) -> u32 {
+  //   self.len
+  // }
+
+  pub(crate) fn has(&self, id: PromiseId) -> bool {
+    self.ring.get(Self::ring_idx(id)).is_some() || self.map.contains_key(&id)
+  }
+
+  pub(crate) fn take(&mut self, id: PromiseId) -> Option<PromiseResolver> {
+    let ring_start = if self.cursor < (RING_SIZE as u32) {
+      0
+    } else {
+      self.cursor - RING_SIZE as u32
+    };
+    let resolver = if id >= ring_start {
+      self.ring.get_mut(Self::ring_idx(id)).unwrap().take()
+    } else {
+      self.map.remove(&id)
+    };
+    if resolver.is_some() {
+      self.len -= 1;
+    }
+    resolver
+  }
+
+  pub(crate) fn head(&self) -> PromiseId {
+    self.cursor as PromiseId
+  }
+
+  pub(crate) fn put(&mut self, resolver: PromiseResolver) -> PromiseId {
+    let id = self.cursor;
+    self.cursor += 1;
```

Comment:
Some of my work touches on a ring buffer of sorts, where the implementation has been done such that the cursor is always traveling backwards. I don't know why that is but I have some faith in the original implementers that they didn't do it randomly but with good reason.

So, maybe give it a minute of thought if that would make sense in this case? (Though the use cases are quite different indeed.)

Reply:
Further thinking on this: The ring buffer that I referred does not have a concept of  IDs in the sense of the PromiseRing, instead it only has the "head" which keeps consuming the tail as it goes. In that scenario the inverted order probably has benefits but for Deno's use case it likely does not make sense since indexing by IDs needs to work etc.